session/auth/authprovider/tokenseed.go

```diff
@@ -0,0 +1,65 @@
+package authprovider
+
+import (
+	"crypto/rand"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+type tokenSeeds struct {
+	mu  sync.Mutex
+	dir string
+}
+
+type seed struct {
+	Seed []byte
+}
+
+func (ts *tokenSeeds) getSeed(host string) ([]byte, error) {
+	ts.mu.Lock()
+	defer ts.mu.Unlock()
+
+	fp := filepath.Join(ts.dir, ".token_seed")
+	dt, err := ioutil.ReadFile(fp)
+	m := map[string]seed{}
+	if err != nil {
+		if !errors.Is(err, os.ErrNotExist) {
+			return nil, err
+		}
+	} else {
+		if err := json.Unmarshal(dt, &m); err != nil {
+			return nil, errors.Wrapf(err, "failed to parse %s", fp)
+		}
+	}
+	v, ok := m[host]
+	if !ok {
+		v = seed{Seed: newSeed()}
+	}
+
+	m[host] = v
+
+	dt, err = json.MarshalIndent(m, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+
+	if err := os.MkdirAll(filepath.Dir(fp), 0701); err != nil {
```

Comment:
Just double checking, we want anyone to have access to paths inside the dir, but not be able to list the dir? Or is this supposed to be `0700`?

Reply:
On my machine it is `0755` actually, but cli does not seem to be consistent. The search bit should be set though as not all things in `.config` are secret. Eg. it may contain plugins that need to be executable.

FYI @thaJeztah

Reply:
In most cases, this would be inside the user's home-directory, which would likely not be accessible to the "world" already, correct? 😅

> but cli does not seem to be consistent

Can you elaborate what's not consistent? Something that needs to be fixed?

Reply:
@thaJeztah 

eg. https://github.com/docker/cli/blob/master/cli/context/store/metadatastore.go#L31
https://github.com/docker/cli/blob/master/cli/config/configfile/file.go#L187

There's at least one example for both cases

Reply:
updated to 0755

Reply:
Opened https://github.com/docker/cli/issues/2727 for tracking on the CLI side.

Would `0755` be problematic if a custom config-directory is specified through `--config` (potentially outside of `~/` ?

Reply:
Not really problematic as the file itself is `0600` and this is just the parent directory. But if you want to be more safe I can put it back to `0700`. In 99% of the cases the directory already exists and all this will be ignored.